Feat/talk2scholars pdf support

ansh-info opened the pull request:
# For authors

## Description
Please:
1. Provide a summary of the modifications made and any associated issue (if applicable).
2. Include relevant context and motivation for the changes.
3. If this relates to a change in any website's frontend, kindly attach a screenshot of the adjustment from your localhost.
4. List any dependencies necessary for implementing this change.

## Fixes # (issue) Mention the issue number.

## Type of change
Please delete options that are not relevant.
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?
Please describe the tests you conducted to verify your changes. These may involve creating new test scripts or updating existing ones.
- [ ] Added new test(s) in the ```tests``` folder
- [ ] Added new function(s) to an existing test(s) (e.g.: ```tests/testX.py```)
- [ ] No new tests added (Please explain the rationale in this case)

## Checklist
- [ ] My code follows the style guidelines mentioned in the Code/DevOps guides
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (e.g. MkDocs)
- [ ] My changes generate no new warnings
- [ ] I have added or updated tests (in the ```tests``` folder) that prove my fix is effective or that my feature works
- [ ] New and existing tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

# For reviewers

## Checklist pre-approval
- [ ] Is there enough documentation?  
- [ ] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
- [ ] Does the test(s) successfully test edge/corner cases?
- [ ] Does the PR pass the tests? (if the repository has continuous integration)

## Checklist post-approval
- [ ] Does this PR merge ```develop``` into ```main```? If so, please make sure to add a prefix (feat/fix/chore) and/or a suffix BREAKING CHANGE (if it's a major release) to your commit message.
- [ ] Does this PR close an issue? If so, please make sure to descriptively close this issue when the PR is merged.

## Checklist post-merge
- [ ] When you approve of the PR, merge and close it (Read this [article](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/about-merge-methods-on-github) to know about different merge methods on GitHub)
- [ ] Did this PR merge ```develop``` into ```main``` and is it suppose to run an automated release workflow (if applicable)? If so, please make sure to check under the "Actions" tab to see if the workflow has been initiated, and return later to verify that it has completed successfully.

gurdeep330 commented on aiagents4pharma/talk2scholars/tools/paper_download/download_arxiv_input.py:
I'd recommend discussing this with @utkarshsawant65 to ensure you both are on the same page.

ansh-info replied:
Sure, we had a conversation regarding the update, thank you

gurdeep330 commented on aiagents4pharma/talk2scholars/tools/pdf/question_and_answer.py:
It's great that you are considering implementing to rank papers by relevance. I have 2 questions/suggestions:
1. Why don't you use an off-the-shelf model for reranking? Based on our previous discussion, please check out NVIDIA's RAG pipeline described [here](https://build.nvidia.com/nvidia/build-an-enterprise-rag-pipeline). If this is not feasible in the current PR, please feel free to create an issue and address it in another PR.
2. Now that we have FAISS implemented, is there any reason to still keep InMemoryVectoreStore?

ansh-info replied:
Thank you, 

1. 
2. Regarding the 2nd point I updated the code and removed the InMemoryVectoreStore, yesterday I had some issues installing faiss on my Mac so kept the above as a fallback, now the code fully supports faiss

ansh-info replied:
I have added support for nims reranker in our pipeline by following this [link](https://build.nvidia.com/nvidia/llama-3_2-nv-embedqa-1b-v2?snippet_tab=Python), and right now I have the model and the api param fetch in the code itself, but I will update in the upcoming pr where I will break the code into helper utilities

ansh-info replied:
I have added the it in hydra config.yaml, just splitting into helper functions is the remaining task.

gurdeep330 commented on app/frontend/streamlit_app_talk2scholars.py:
I suggest extending the functionality from single-PDF upload to multi-PDF upload. Feel free to open a separate issue and submit a PR later to implement this if you prefer. Thx

ansh-info replied:
Sure, after this update we ca upload multiple pdfs

ansh-info replied:
Hi, add the support for multi pdf, please have a look 
<img width="1222" alt="image" src="https://github.com/user-attachments/assets/425ebe77-85a4-4034-b911-a2f0ddca72e2" />

gurdeep330 replied:
@ansh-info 

Unfortunately, it still doesn't work for me 😢 

![image](https://github.com/user-attachments/assets/de03a20d-f168-4725-8015-338337b2adb6)

gurdeep330 replied:
@ansh-info, even after installing FAISS and having already read from Zotero, I am unable to get to do RAG. Could you please describe the steps to install all the libraries?

And the answer to your question: To update the state, either follow the [tutorial from LangGraph](https://langchain-ai.github.io/langgraph/how-tos/human_in_the_loop/edit-graph-state/#agent), or follow my code [here](https://github.com/VirtualPatientEngine/AIAgents4Pharma/blob/e1629cd0e4cd3fd91ef5fe8e9e08dc717a7fd71a/app/frontend/streamlit_app_talk2scholars.py#L243) from lines 243 to 253.

gurdeep330 reviewed (changes requested):
@ansh-info Good work implementing FAISS and adding multi-pdf support😃 

I have a few comments below. Let me know if you have questions!

ansh-info commented on aiagents4pharma/talk2scholars/state/state_talk2scholars.py:
```diff
@@ -45,28 +46,27 @@ def replace_dict(existing: Dict[str, Any], new: Dict[str, Any]) -> Dict[str, Any
         papers (Dict[str, Any]): Stores the research papers retrieved from the agent's queries.
         multi_papers (Dict[str, Any]): Stores multiple recommended papers from various sources.
-        zotero_read (Dict[str, Any]): Stores the papers retrieved from Zotero.
+        article_data (Dict[str, Any]): Stores the papers retrieved from Zotero with their metadata.
```
I have updated it, Thank you

dmccloskey commented on aiagents4pharma/talk2scholars/tools/pdf/question_and_answer.py:
```diff
@@ -50,168 +41,605 @@ class QuestionAndAnswerInput(BaseModel):
+        answer_text = result.get("output_text", "No answer generated.")
+
+        # Get paper titles for sources
+        paper_titles = {}
```
Would be it be possible to instead use Zotero to create the citation? As a reference manager, Zotero is able to generate properly formatted citations. Also, I am missing the connection between the `vector_store` and Zotero (i.e., the `document_store`). How I would imagine the pipeline would be the following:

1. Populate Zotero with the relevant PDF Documents for a collection
2. Chunk the documents in the collection 
3. User asks question about documents in the collection
4. Perform RAG and return the results with citations generated by Zotero

The benefit of step 1 is that it does not matter which tool downloads the PDFs. It could even be the user that downloads the PDFs and uploads them to their Zotero account. 

Note that step 2 links FAISS indices with Zotero collections. The idea being that aa Zotero account could potentially have multiple collections on different topics that the user would like to search over separately. If the user wants to search over their entire library then we would need an index that includes all of their collections.

ansh-info replied:
Thank you fir the suggestion, I will have to look how we could use zotero as the reference manager, but regarding the `vector_store` or `document_store`, its the same I just named it document_store. The data from zoteor is saved under `article_data`.

I will come back to this comment again after I develop a better understanding, but yes I see the clear picture, both regarding the working of pipeline and the steps you mentioned

ansh-info replied:
Hi,
So far, we’ve implemented PDF reading from Zotero, FAISS support, and the Nvidia reranker. We can handle the first three bullet points you mentioned. For the last one, I still need to add support. Please correct me if I’m wrong, but here’s how I’ve understood it:
When an answer is generated, we should include a citation at the bottom (i.e., title, author, and source PDF). To do this, right before returning the answer, we’ll call the Zotero read helper. Using the paper_id, we’ll fetch the citation for the selected paper. We’ll need to write a new function to handle this citation logic. I plan to implement this in future PRs after I understand how Zotero handles citations.
Regarding your second point:
Right now, we also generate the vector store on the fly and assign it a unique ID. So, indexing the entire collection is possible—we just need to look into how embeddings are handled for that specific index.
Let me know if I misunderstood anything. I’ll also explore Zotero’s API for citation support.

dmccloskey replied:
Hi Ansh,

That sounds about right. 

Re 4: By citation I am referring to the Zotero citation/bibliography https://www.zotero.org/support/creating_bibliographies. I believe the `copy to clipboard` is what we are looking for since it generates a formatted text that can be injected into the response.

ansh-info replied:
Thank you for sharing the link I have created an issue #161 -> which will be covered in the next pr's

dmccloskey commented on app/frontend/streamlit_app_talk2scholars.py:
```diff
@@ -131,19 +131,44 @@ def process_pdf_upload():
-            "arxiv_id": None,  # placeholder for an arXiv id if applicable
+
+        # Create a unique ID for the uploaded PDF
+        pdf_id = f"uploaded_{pdf_file.name.replace(' ', '_').replace('.', '_')}"
```
I would recommend adding either a random index or a time-stamp to ensure uniqness.

ansh-info replied:
Sure, now we are using time-stamp to each PDF’s unique ID

dmccloskey reviewed (changes requested):
Very nice start to this very difficult and technically challenging task 👏. Please let me know if you have any questions on my comments. I suggest we have a meeting to discuss the Zotero integration part when @gurdeep330 is back from vacation next week.

ansh-info commented on aiagents4pharma/talk2scholars/tools/paper_download/abstract_downloader.py:
Yes, if you now look at the `download_arxiv_input` I have simplified the code to just fetch the pdf url, and introduced a standard data structure, as we no longer need binary data and we just ned the pdf_url and pypdfloader will take care of it, regarding the int the `__init__ ` I just removed the unnecessary imports

ansh-info commented on aiagents4pharma/talk2scholars/tools/pdf/question_and_answer.py:
```diff
@@ -50,168 +41,605 @@ class QuestionAndAnswerInput(BaseModel):
+            return results
+
+    @staticmethod
+    def _cosine_similarity(
```
Hi, we are using built in similarity scores now, I also took help from the link you shared regarding the rah pipeline and now we use nims reranker instead of custom reranker that I have implemented earlier

gurdeep330 commented:
@ansh-info., unfortunately, I came across some issues while testing the application:
1. Even after the paper is downloaded, the agent is unable to perform RAG.
![image](https://github.com/user-attachments/assets/1057d490-f18d-4283-9ca5-3a7c25904c21)
2. While the agent can retrieve all articles from my Zotero library, it is unable to perform RAG over the articles.
![image](https://github.com/user-attachments/assets/9ea68171-b473-40a3-b8c4-3de6f965bcce)

Do you think there needs to be an update to requirements.txt?

gurdeep330 reviewed (changes requested):
@ansh-info., unfortunately, I came across some issues while testing the application:
1. Even after the paper is downloaded, the agent is unable to perform RAG.
![image](https://github.com/user-attachments/assets/1057d490-f18d-4283-9ca5-3a7c25904c21)
2. While the agent can retrieve all articles from my Zotero library, it is unable to perform RAG over the articles.
![image](https://github.com/user-attachments/assets/9ea68171-b473-40a3-b8c4-3de6f965bcce)

Perhaps you need to update requirements.txt?

ansh-info commented:
> @ansh-info., unfortunately, I came across some issues while testing the application:
> 
> 1. Even after the paper is downloaded, the agent is unable to perform RAG.
>    ![image](https://private-user-images.githubusercontent.com/43615259/431375548-1057d490-f18d-4283-9ca5-3a7c25904c21.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQxMTg5NTYsIm5iZiI6MTc0NDExODY1NiwicGF0aCI6Ii80MzYxNTI1OS80MzEzNzU1NDgtMTA1N2Q0OTAtZjE4ZC00MjgzLTljYTUtM2E3YzI1OTA0YzIxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDA4VDEzMjQxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE1MGI4OWIyYWYwNzIwNjg2NjhhNmUwY2EyNDMxNTdiMmNmZWQyNjY0ZTk5Yzg1NTUyN2NmZDRmZGY2MzZjYzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.P8LfRIDHLpDOO-2Ecb1lRs9c4RW6KwIt1Cw-1-jLVT0)
> 2. While the agent can retrieve all articles from my Zotero library, it is unable to perform RAG over the articles.
>    ![image](https://private-user-images.githubusercontent.com/43615259/431379571-9ea68171-b473-40a3-b8c4-3de6f965bcce.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQxMTg5NTYsIm5iZiI6MTc0NDExODY1NiwicGF0aCI6Ii80MzYxNTI1OS80MzEzNzk1NzEtOWVhNjgxNzEtYjQ3My00MGEzLWI4YzQtM2RlNmY5NjViY2NlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDA4VDEzMjQxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI2ZjUyMzY3MjM1YWU1MTkxOWYyYjk5NWRiODgzZTZiODUxY2JkYjVkNWIwZDRmMGZiMjk5NmQ3NjJmZGVlNDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.s-Qoevr9F_u56gtBj5Ncl2xf8sB5DVkQp_hM2aR87RM)
> 
> Perhaps you need to update requirements.txt?

I the only way to install FAISS is through conda so will updating the requirements.txt might not work? please check the [link](https://github.com/facebookresearch/faiss/blob/main/INSTALL.md)

gurdeep330 commented:
> > @ansh-info., unfortunately, I came across some issues while testing the application:
> > 
> > 1. Even after the paper is downloaded, the agent is unable to perform RAG.
> >    ![image](https://private-user-images.githubusercontent.com/43615259/431375548-1057d490-f18d-4283-9ca5-3a7c25904c21.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQxMTg5NTYsIm5iZiI6MTc0NDExODY1NiwicGF0aCI6Ii80MzYxNTI1OS80MzEzNzU1NDgtMTA1N2Q0OTAtZjE4ZC00MjgzLTljYTUtM2E3YzI1OTA0YzIxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDA4VDEzMjQxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE1MGI4OWIyYWYwNzIwNjg2NjhhNmUwY2EyNDMxNTdiMmNmZWQyNjY0ZTk5Yzg1NTUyN2NmZDRmZGY2MzZjYzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.P8LfRIDHLpDOO-2Ecb1lRs9c4RW6KwIt1Cw-1-jLVT0)
> > 2. While the agent can retrieve all articles from my Zotero library, it is unable to perform RAG over the articles.
> >    ![image](https://private-user-images.githubusercontent.com/43615259/431379571-9ea68171-b473-40a3-b8c4-3de6f965bcce.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQxMTg5NTYsIm5iZiI6MTc0NDExODY1NiwicGF0aCI6Ii80MzYxNTI1OS80MzEzNzk1NzEtOWVhNjgxNzEtYjQ3My00MGEzLWI4YzQtM2RlNmY5NjViY2NlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDA4VDEzMjQxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI2ZjUyMzY3MjM1YWU1MTkxOWYyYjk5NWRiODgzZTZiODUxY2JkYjVkNWIwZDRmMGZiMjk5NmQ3NjJmZGVlNDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.s-Qoevr9F_u56gtBj5Ncl2xf8sB5DVkQp_hM2aR87RM)
> > 
> > Perhaps you need to update requirements.txt?
> 
> This could be due to FAISS not installed ? The recommended way to install FAISS is through conda so updating the requirements.txt might not work? thier installation guide [link](https://github.com/facebookresearch/faiss/blob/main/INSTALL.md), I think I can add it in the readme?

Hi @ansh-info 

Yes, I'd recommend updating the README. However, even after installing FAISS, I see that it answers questions using the abstract and not the PDF of the article (see below). In this case, my Zotero account has 2 articles with their PDFs attached. While the agent was able to retrieve their records from Zotero, it still doesn't answer using the PDF. I can say so because, in the BioBridge paper, the authors used the models ESM1b and UniMol to encode protein sequences and drug molecules, resp.
![image](https://github.com/user-attachments/assets/50bfaef5-54b4-4689-ae60-5677b3c61290)

dmccloskey commented on aiagents4pharma/talk2scholars/tools/pdf/question_and_answer.py:
```diff
@@ -50,168 +56,544 @@ class QuestionAndAnswerInput(BaseModel):
+    # Choose papers to use
+    selected_paper_ids = []
+
+    if paper_ids:
```
All three scenarios are covered by the tests (which is great 👍) but I don't fully understand why there are these three branches. Specifically, I see two options: 1) list of specific papers, and 2) use of the reranker. If my understanding is correct, there would be 4 total branches i.e., list of papers without reranker, list of papers with reranker, all papers without reranker, and all papers with reranker.

ansh-info replied:
Thank you for your detailed review! I appreciate you noticing the coverage of all three scenarios in our tests.

Let me clarify the branching logic in our paper selection process:

Our code has three distinct scenarios that address specific user needs:

1. **User specifies paper IDs**: When users know exactly which papers they want (via the `paper_ids` parameter), we use precisely those papers without any ranking. This is a direct selection scenario.

2. **Use all available papers**: This happens either when:
   - The user explicitly requests all papers (`use_all_papers=True`)
   - The user has uploaded custom papers (`has_uploaded_papers=True`)
   
   In both cases, we don't need any ranking since we're including everything.

3. **Automatic selection via reranker**: When the user doesn't specify which papers to use and doesn't request all papers, we use the NVIDIA reranker to automatically find the most relevant papers for their question.

The reason we don't have four branches (as you suggested with combinations of specific/all papers with/without reranker) is that the reranker is only useful when we need to automatically select a subset of papers. When papers are already specified (scenario 1) or when all papers should be used (scenario 2), applying the reranker would be unnecessary processing.

In other words:
- If papers are explicitly specified, ranking them doesn't change which ones we use
- If all papers should be used, ranking them doesn't change which ones we use
- The reranker is only valuable when making an automatic selection of the most relevant papers

I hope this explains our approach! Let me know if you have any other questions.

dmccloskey replied:
@ansh-info Thanks for the explanation. Make sense 👍

dmccloskey reviewed (changes requested):
Nice work implementing the reranker 😊. I have a few comments on the use of the logger in the tests and the code logic when implementing the reranker in the Q&A tool.

ansh-info commented on aiagents4pharma/talk2scholars/tools/pdf/question_and_answer.py:
```diff
@@ -50,168 +56,544 @@ class QuestionAndAnswerInput(BaseModel):
+            logger.info("Retrieved %d chunks using MMR", len(results))
+            return results
+
+        return []
```
Good observation! Earlier, I had a fallback to standard similarity in case MMR failed, but I removed that in a previous update. So now, there's no reason to keep the `use_mmr` condition. I've simplified the function by removing the `use_mmr` parameter and making MMR the only retrieval method. This also removes the empty return case and ensures we always get diverse and relevant results. It was my mistake—I forgot to remove the `if` condition for `use_mmr` earlier.

dmccloskey replied:
Sounds good 👍

ansh-info commented on aiagents4pharma/talk2scholars/tools/pdf/question_and_answer.py:
```diff
@@ -50,168 +56,544 @@ class QuestionAndAnswerInput(BaseModel):
+        Returns:
+            List of document chunks
+        """
+        if not self.vector_store:
```
You're right - there's redundant checking and building of the vector store. I'll centralize this logic in the question_and_answer_tool function.

ansh-info commented on aiagents4pharma/talk2scholars/tools/pdf/question_and_answer.py:
```diff
@@ -50,168 +56,544 @@ class QuestionAndAnswerInput(BaseModel):
+
+    # Retrieve relevant chunks across selected papers
+    relevant_chunks = vector_store.retrieve_relevant_chunks(
+        query=question, paper_ids=selected_paper_ids, top_k=10, use_mmr=True
```
Sure, I forgot to add it

ansh-info commented on aiagents4pharma/talk2scholars/tests/test_question_and_answer_tool.py:
```diff
@@ -2,249 +2,829 @@
+                )
+
+        # Check if the appropriate log message is present
+        self.assertIn("Error loading paper test_paper: Loading error", log.output[0])
```
Thank you for your valuable feedback. I agree that we should focus on testing the observable behavior—such as verifying the returned error messages and state changes—instead of asserting on log outputs. This approach helps keep our tests focused on the core business logic and makes them more maintainable. I’ve updated the tests accordingly. Thanks again for your positive remarks on our test coverage!

dmccloskey commented on aiagents4pharma/talk2scholars/tools/pdf/question_and_answer.py:
```diff
@@ -50,168 +56,544 @@ class QuestionAndAnswerInput(BaseModel):
+        # Here we assign a dummy score based on rank position;
+        # the actual re-ranker sorts the documents.
+        ranked_papers = [
+            (doc.metadata["paper_id"], float(top_k - idx))
```
Thanks for the explanation 🙂

ansh-info commented:
Hi @dmccloskey and @gurdeep330, I noticed that for embeddings
I was Looping over `embed_documents([chunk])`(calling embed_documents() inside a loop, one document chunk at a time.) which was inefficient. Every call triggered a separate API call. I just now updated the code to proceed embedding in a batch, which in results made the embedding almost instant, I would like you to please review it as well, and i have updated the test for it, sorry for using the wrong method initially....

Plese check the updated code at line number 254 `def retrieve_relevant_chunks`

dmccloskey commented on aiagents4pharma/talk2scholars/install.md:
```diff
@@ -0,0 +1,66 @@
+
+Follow the official FAISS installation instructions [here](https://github.com/facebookresearch/faiss/blob/main/INSTALL.md)
+
+Make sure your system has a compatible CUDA toolkit version if you are using the GPU variant.
```
I would recommend omitting this line. There are some additional nuances (e.g., gpu only for linux machines).

ansh-info replied:
Sure, I updated it.

dmccloskey reviewed (approved):
Thanks for addressing my comments 🙏. The updated tests look good.

dmccloskey commented:
> Hi @dmccloskey and @gurdeep330, I noticed that for generating embeddings, I was Looping over `embed_documents([chunk])` (calling embed_documents() inside a loop, one document chunk at a time.) which was inefficient. Every call triggered a separate API call.
> 
> I just now updated the code to proceed embedding in a batch, which in results made the embedding almost instant, I would like you to please review it as well, and i have updated the test for it, sorry for using the wrong method initially....
> 
> Plese check the updated code at line number 254 `def retrieve_relevant_chunks`

Hi @ansh-info I completely missed this 😅 so I am glad you spotted it 👍.

gurdeep330 reviewed (changes requested):
Hi @ansh-info,

Thanks for fixing the code:

1. I am still facing the issue of not being able to ask questions to the articles stored in my Zotero account:
![image](https://github.com/user-attachments/assets/e88bc4a6-9d22-4de8-b017-200bd75bd693)
![image](https://github.com/user-attachments/assets/26154ca7-6ae9-4d27-a353-1ed9f154e758)


2. I found a minor bug:
![image](https://github.com/user-attachments/assets/a1c19c4d-d346-4dd9-8c58-c4bacb23d782)
Though it had indeed downloaded the correct paper (BioBridge) because it could correclty answer questions related to the downloaded paper.

ansh-info commented:
> Hi @ansh-info,
> 
> Thanks for fixing the code:
> 
> 1. I am still facing the issue of not being able to ask questions to the articles stored in my Zotero account:
>    ![image](https://private-user-images.githubusercontent.com/43615259/436075988-e88bc4a6-9d22-4de8-b017-200bd75bd693.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU3MzgwMjYsIm5iZiI6MTc0NTczNzcyNiwicGF0aCI6Ii80MzYxNTI1OS80MzYwNzU5ODgtZTg4YmM0YTYtOWQyMi00ZGU4LWIwMTctMjAwYmQ3NWJkNjkzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDI3VDA3MDg0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWEyOThkZWRiNTljYzJmYWE2M2VkODgzNTE3MzhjZDk5M2FiNmJlYWI1Yjg3OGUxYWRkYzA5YmQ2NGM0ZjczMTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.czXnz-0Kv25HmOgELidPn4Rr4Z-3Gz8hwI6OZchJUX4)
>    ![image](https://private-user-images.githubusercontent.com/43615259/436076138-26154ca7-6ae9-4d27-a353-1ed9f154e758.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU3MzgwMjYsIm5iZiI6MTc0NTczNzcyNiwicGF0aCI6Ii80MzYxNTI1OS80MzYwNzYxMzgtMjYxNTRjYTctNmFlOS00ZDI3LWEzNTMtMWVkOWYxNTRlNzU4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDI3VDA3MDg0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQyOWJhOWJmMmQyODA3ZWUxZjU4NjQyZmIzNzI1ZjM1ZDU2ZWMxMmQzYTk3NmYzOGZjNTJlNWY3ZTY3NmZhNmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.q9qHNPeY0J_kNYkj1fGkY1F8PzJmKtM1ELvMilU8BuI)
> 2. I found a minor bug:
>    ![image](https://private-user-images.githubusercontent.com/43615259/436072086-a1c19c4d-d346-4dd9-8c58-c4bacb23d782.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU3MzgwMjYsIm5iZiI6MTc0NTczNzcyNiwicGF0aCI6Ii80MzYxNTI1OS80MzYwNzIwODYtYTFjMTljNGQtZDM0Ni00ZGQ5LThjNTgtYzRiYWNiMjNkNzgyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDI3VDA3MDg0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ0OTg1Njc2ZDMzODc5MjFiNDg2YWZmNzRmZTcyNDA2ODc1NDRmYmI2MjUwMWE2OTIxODlmZDg3MDIyMzNlZWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.9yHZ5_NYVTS5Pqy80xRwh-lbzwTXDQIT7-_ZRZuEF00)
>    Though it had indeed downloaded the correct paper (BioBridge) because it could correclty answer questions related to the downloaded paper.

Hi @gurdeep330 ,
Sorry to hear you're still facing this issue. I sincerely apologize for the late response — I got a bit busy with my upcoming semester exams.
I will take a closer look to debug what's happening and work on fixing it. Thank you for your patience!

gurdeep330 commented:
@ansh-info thanks for the update.
I tested your updated PR but I am still facing some issues:
![image](https://github.com/user-attachments/assets/5388bbc6-a662-454c-8924-9a306250cdff)
